### PR TITLE
PSEC-666 Upload Scan CLI and Maven plugin

### PR DESCRIPTION
**Pull request by madisoncool**
- Updating to ignore IntelliJ files
- Adding CLI and Maven plugin
- Updating gitignore for target and non-blank files

### CODE REVIEW

**Comment by madisoncool** [comment]
A couple of open questions:
- How long should we retry to get Watchtower results?
- For the Maven plugin, should this be a part of the compile stage?

File: watchtower-cli/pom.xml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		<dependency>
+			<groupId>com.github.tomakehurst</groupId>
+			<artifactId>wiremock-jre8</artifactId>
+			<version>2.27.2</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>com.tracelink.appsec</groupId>
+			<artifactId>watchtower-test</artifactId>
+			<version>2.3.0-SNAPSHOT</version>
```
Comment (tophersmith) [inline watchtower-cli/pom.xml]: should probably be ${project.version]

File: watchtower-cli/src/main/java/com/tracelink/appsec/watchtower/cli/reporter/AbstractReporter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Abstract class to report Watchtower scan results in various formats. Reports a summary of the
+ * scan results to the console.
+ *
+ * @author mcool
+ */
+public abstract class AbstractReporter {
+
+	private final Logger log;
```
Comment (tophersmith) [inline watchtower-cli/src/main/java/com/tracelink/appsec/watchtower/cli/reporter/AbstractReporter.java]: I think you can do Logger.getClass(this.getClass()) and it should get the concrete classname

File: watchtower-maven-plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	<dependencies>
+		<dependency>
+			<groupId>com.tracelink.appsec</groupId>
+			<artifactId>watchtower-cli</artifactId>
+			<version>${project.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>com.tracelink.appsec</groupId>
+			<artifactId>watchtower-test</artifactId>
+			<version>2.3.0-SNAPSHOT</version>
```
Comment (tophersmith) [inline watchtower-maven-plugin/pom.xml]: ditto on project.version

**Review by tophersmith — COMMENTED** [review]
Looking good. few comments to sort out. one thought: should we come up with a way for the cli/maven plugin to check the version of watchtower itself to make sure it's compatible? I don't know exactly how that'd work out...
Also there's a merge conflict on the README.md file

File: watchtower-cli/src/main/java/com/tracelink/appsec/watchtower/cli/executor/UploadScanExecutor.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+	/**
+	 * Zips the given target file or directory and returns the path to the zipped file.
+	 *
+	 * @param target the file or directory to zip
+	 * @return path to the zipped file
+	 * @throws IllegalArgumentException if the target directory or file does not exist
+	 * @throws IOException              if there is a problem zipping the target
+	 */
+	private Path zipTarget(String target) throws IllegalArgumentException, IOException {
```
Comment (tophersmith) [inline watchtower-cli/src/main/java/com/tracelink/appsec/watchtower/cli/executor/UploadScanExecutor.java:68]: watchtower also has a zip4j library already in it, so you could use that here too

**Comment by madisoncool** [comment]
Yes--saw the merge conflict and will fix it along with the comments you had. I thought about the version compatibility issue but I don't know how easy it would be to maintain an endpoint to check compatibility. Figured it would be best just to include in the release notes if a new version of Watchtower Core is no longer compatible with older versions of the CLI/maven plugin. If we change the REST API significantly we have to update and release the CLI/maven plugin simultaneously anyway. Might be worth it to add something like that down the road but not sure we need it now.

**Comment by tophersmith** [comment]
Yeah no problem. Likely when we put the link between watchtower and synapse, we'll have to work on a versioning scheme of some kind anyway

**Review by tophersmith — APPROVED** [review]

**Pull request merged**